onnxruntime/core/session/inference_session.cc

```diff
@@ -97,37 +98,134 @@ inline std::basic_string<T> GetCurrentTimeString() {
 
 std::atomic<uint32_t> InferenceSession::global_session_id_{1};
 
-InferenceSession::InferenceSession(const SessionOptions& session_options,
-                                   logging::LoggingManager* logging_manager)
-    : session_options_(session_options),
-      graph_transformation_mgr_(session_options.max_num_graph_transformation_steps),
-      logging_manager_(logging_manager),
-      thread_pool_(concurrency::CreateThreadPool("intra_op_thread_pool",
-                                                 session_options.intra_op_num_threads)),
-      inter_op_thread_pool_(session_options.execution_mode == ExecutionMode::ORT_PARALLEL
-                                ? concurrency::CreateThreadPool("inter_op_thread_pool",
-                                                                session_options.inter_op_num_threads)
-                                : nullptr),
-      session_state_(execution_providers_,
-                     session_options.enable_mem_pattern && session_options.execution_mode == ExecutionMode::ORT_SEQUENTIAL,
-                     thread_pool_.get(),
-                     inter_op_thread_pool_.get()),
-      insert_cast_transformer_("CastFloat16Transformer") {
+Status InferenceSession::FinalizeSessionOptions(SessionOptions& session_options,
+                                                const ONNX_NAMESPACE::ModelProto& model_proto) {
+  InferenceSessionUtils inference_session_utils(logging::LoggingManager ::DefaultLogger());
+
+  ORT_RETURN_IF_ERROR(inference_session_utils.ParseOrtConfigJsonInModelProto(model_proto));
+
+  ORT_RETURN_IF_ERROR(inference_session_utils.ParseSessionOptionsFromModelProto(session_options));
+
+  return Status::OK();
+}
+
+void InferenceSession::ConstructorCommon(const SessionOptions& session_options,
+                                         logging::LoggingManager* logging_manager) {
   ORT_ENFORCE(Environment::IsInitialized(),
               "Environment must be initialized before creating an InferenceSession.");
+
+  const SessionOptions* final_session_options;
+  // the model is to be checked for an ORT config json that may hold some/all session options
+  if (session_options.check_model_for_ort_config) {
+    // In theory should not throw on this line unless this internal class' APIs are being called incorrectly.
+    // It is a good sanity check to enforce this
+    ORT_ENFORCE(model_proto_, "Need model to be provided to check for ORT config within it");
+
+    SessionOptions constructed_session_options;
+
+    auto status = FinalizeSessionOptions(constructed_session_options, *model_proto_);
+    ORT_ENFORCE(status.IsOK(), "Could not finalize session options while constructing the inference session. Error Message: ",
+                status.ErrorMessage());
+
+    final_session_options = &constructed_session_options;
+  } else {
```

Comment:
constructed_session_options will get destructed at the end of the if block and the final_session_options ptr will be dangling.

Reply:
Thanks for catching. Strangely, this didn't even invoke a compiler warning...I ll fix it.

Reply:
Should be fixed now. Moved the declaration to outside the `if`.